airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py

```diff
@@ -0,0 +1,227 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+import json
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Union
+from urllib.parse import parse_qsl, urlparse
+
+import pendulum
+import requests
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth.core import HttpAuthenticator
+from requests.auth import AuthBase
+
+
+class LinnworksStream(HttpStream, ABC):
+    http_method = "POST"
+
+    def __init__(self, authenticator: Union[AuthBase, HttpAuthenticator] = None, start_date: str = None):
+        super().__init__(authenticator=authenticator)
+
+        self._authenticator = authenticator
+        self.start_date = start_date
+
+    @property
+    def url_base(self) -> str:
+        return self._authenticator.get_server()
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        if response.status_code != requests.codes.ok:
+            return None
+
+        json = response.json()
+        if not isinstance(json, list):
+            json = [json]
+        for record in json:
+            yield record
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        delay_time = response.headers.get("Retry-After")
+        if delay_time:
+            return int(delay_time)
+
+
+class LinnworksGenericPagedResult(ABC):
+    # https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Generic-GenericPagedResult
+    @abstractmethod
+    def paged_result(self, response: requests.Response) -> Mapping[str, Any]:
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        result = self.paged_result(response)
+
+        if result["PageNumber"] < result["TotalPages"]:
+            return {
+                "PageNumber": result["PageNumber"] + 1,
+                "EntriesPerPage": result["EntriesPerPage"],
+                "TotalEntries": result["TotalEntries"],
+                "TotalPages": result["TotalPages"],
+            }
+
+
+class Location(LinnworksStream):
```

Comment:
I don't see how it can benefit from caching. These two streams perform exactly N+1 requests. The documentation you pointed to talks about reducing N+N requests to N+1.

Stream `StockLocations` requests `/api/Inventory/GetStockLocations` once. Then iterates over it's response and invokes stream `Location` which requests `/api/Locations/GetLocation?pkStockLocationId={StockLocationId}`. Each value of `StockLocationId` is unique.